cdap-metrics-sidecar/src/main/java/io/cdap/cdap/app/JMXMetricsCollector.java

```diff
@@ -0,0 +1,134 @@
+  protected void startUp() {
+    LOG.info("Starting JMXMetricsCollector.");
+  }
+
+  @Override
+  protected void shutDown() {
+    LOG.info("Shutting down JMXMetricsCollector has completed.");
+  }
+
+  @Override
+  protected void runOneIteration() {
+    // TODO: Get component name according to the pod this service is running in
+    MetricsContext context = metricsCollector.getContext(Constants.Metrics.APP_FABRIC_CONTEXT);
+    MBeanServerConnection mBeanConn;
+    try {
+      JMXServiceURL serviceUrl = new JMXServiceURL(serverUrl);
+      JMXConnector jmxConnector = JMXConnectorFactory.connect(serviceUrl, null);
+      mBeanConn = jmxConnector.getMBeanServerConnection();
+    } catch (Exception e) {
+      LOG.warn("Error occurred while connecting to JMX server: " + e.getMessage());
+      return;
+    }
+    getAndPublishMemoryMetrics(mBeanConn, context);
+    getAndPublishCPUMetrics(mBeanConn, context);
+    getAndPublishThreadMetrics(mBeanConn, context);
+  }
+
+  private void getAndPublishMemoryMetrics(MBeanServerConnection mBeanConn, MetricsContext context) {
+    MemoryMXBean mxBean;
+    try {
+      mxBean = ManagementFactory.newPlatformMXBeanProxy(mBeanConn, ManagementFactory.MEMORY_MXBEAN_NAME,
+                                MemoryMXBean.class);
+    } catch (IOException e) {
+      LOG.warn("Error occurred while collecting memory metrics from JMX: " + e.getMessage());
+      return;
+    }
+    MemoryUsage heapMemoryUsage = mxBean.getHeapMemoryUsage();
+    context.gauge(Constants.Metrics.JVMResource.HEAP_MEMORY_USED_MB, heapMemoryUsage.getUsed() / megaByte);
+    context.gauge(Constants.Metrics.JVMResource.HEAP_MEMORY_MAX_MB, heapMemoryUsage.getMax() / megaByte);
+  }
+
+  private void getAndPublishCPUMetrics(MBeanServerConnection conn, MetricsContext context) {
+    OperatingSystemMXBean mxBean;
+    try {
+      mxBean = ManagementFactory.newPlatformMXBeanProxy(conn, ManagementFactory.OPERATING_SYSTEM_MXBEAN_NAME,
+                                OperatingSystemMXBean.class);
+    } catch (IOException e) {
+      LOG.warn("Error occurred while collecting memory metrics from JMX: " + e.getMessage());
+      return;
+    }
+    double systemCPULoad = mxBean.getSystemCpuLoad();
+    if (systemCPULoad < 0) {
+      LOG.info("CPU load fro JVM process is not yet available");
+      return;
+    }
+    context.gauge(Constants.Metrics.JVMResource.PROCESS_CPU_LOAD_PERCENT, (long) systemCPULoad * 100);
+  }
+
+  private void getAndPublishThreadMetrics(MBeanServerConnection conn, MetricsContext context) {
+    ThreadMXBean mxBean;
+    try {
+      mxBean = ManagementFactory.newPlatformMXBeanProxy(conn, ManagementFactory.THREAD_MXBEAN_NAME,
+                                ThreadMXBean.class);
+    } catch (IOException e) {
+      LOG.warn("Error occurred while collecting thread metrics from JMX: " + e.getMessage());
+      return;
+    }
+    context.gauge(Constants.Metrics.JVMResource.THREAD_COUNT, mxBean.getThreadCount());
```

Comment:
Any other metric that we can publish here that will be useful? For eg: https://github.com/cdapio/cdap/blob/163ade54065ecac3a4fd80fe3d76a5aee38474f9/cdap-standalone/src/main/java/org/apache/hadoop/util/ReflectionUtils.java#L147 extracts lot of information. We can finalize on a few that will be helpful.